Integration test orchestrator V2 (#1511)
u-root/u-root: integration/README.md

These are VM based tests for core u-root functionality such as:

* retrieving and kexec'ing a Linux kernel,
* DHCP client tests,
* uinit (user init), and
* running unit tests requiring root privileges.

All tests are in the integration/ directory. Within that, there are a

> few subdirectories:

Reviewer: What is the reasoning behind splitting the tests into two subdirectories?

Author: I initially thought I'd modify the tests so that everything in the generic-tests directory could just use the generic initramfs. Later since I was keeping the option for a custom uinit I didn't end up doing that but kept the different directories in case going forward, we wanted to separate tests eg. boot vs. generic vs. go tests. Its not necessary though

* generic-tests/ : most tests can be put under this.

> * golang-tests/ : this is for Go unit tests that can be run inside the VM.

Reviewer: ```suggestion
* gotests/ : this is for Go unit tests that can be run inside the VM.
```

our given kernel and initramfs, and runs the uinit or commands that we want to
test.

Testing mainly relies on 2 packages: [pkg/vmtest](/pkg/vmtest) and
[pkg/qemu](/pkg/qemu).

pkg/vmtest takes in integration test options, and given those and the
environment variables, uses pkg/qemu to start a QEMU VM with the correct command
line and configuration.

The test architecture, kernel and QEMU binary are set using environment
variables.

> The initramfs can come from one of the following ways:

Reviewer: ```suggestion
The initramfs can come from the following sources:
```

> * Custom initramfs: run the tests with an initramfs provided by the user

Reviewer: ```suggestion
* User overridden: when the `UROOT_INITRAMFS` environment variable is used to override the initramfs. The user is responsible for ensuring the initramfs contains the correct binaries.
```

environment variables. You can see the Dockerfile and the config file used to
build the kernel for each supported architecture [here](/.circleci/images).

2. **Run Tests**

> Recall that there are 2 subdirectories with tests, generic-tests/ and gotests/.

Reviewer: Could you also not run `go test ./...` to test both directories?

Author: Yes, it seems to be a bit slower, so updated it accordingly

> Enter each directory and run

Reviewer: ```suggestion
Enter each directory and run:
```
